src/Features/Lsif/Generator/Generator.cs

```diff
@@ -40,23 +42,27 @@ public void GenerateForCompilation(Compilation compilation, string projectPath,
             var topLevelSymbolsWriter = new BatchingLsifJsonWriter(_lsifJsonWriter);
             var topLevelSymbolsResultSetTracker = new SymbolHoldingResultSetTracker(topLevelSymbolsWriter, compilation, _idFactory);
 
-            Parallel.ForEach(compilation.SyntaxTrees, syntaxTree =>
+            var tasks = ArrayBuilder<Task>.GetInstance(project.DocumentIds.Count);
+            foreach (var document in project.Documents)
             {
-                var semanticModel = compilation.GetSemanticModel(syntaxTree);
-
-                // We generate the document contents into an in-memory copy, and then write that out at once at the end. This
-                // allows us to collect everything and avoid a lot of fine-grained contention on the write to the single
-                // LSIF file. Becasue of the rule that vertices must be written before they're used by an edge, we'll flush any top-
-                // level symbol result sets made first, since the document contents will point to that. Parallel calls to CopyAndEmpty
-                // are allowed and might flush other unrelated stuff at the same time, but there's no harm -- the "causality" ordering
-                // is preserved.
-                var documentWriter = new BatchingLsifJsonWriter(_lsifJsonWriter);
-                var documentId = GenerateForDocument(semanticModel, languageServices, topLevelSymbolsResultSetTracker, documentWriter, _idFactory);
-                topLevelSymbolsWriter.FlushToUnderlyingAndEmpty();
-                documentWriter.FlushToUnderlyingAndEmpty();
-
-                documentIds.Add(documentId);
-            });
+                tasks.Add(Task.Run(async () =>
+                {
+                    // We generate the document contents into an in-memory copy, and then write that out at once at the end. This
+                    // allows us to collect everything and avoid a lot of fine-grained contention on the write to the single
+                    // LSIF file. Because of the rule that vertices must be written before they're used by an edge, we'll flush any top-
+                    // level symbol result sets made first, since the document contents will point to that. Parallel calls to CopyAndEmpty
+                    // are allowed and might flush other unrelated stuff at the same time, but there's no harm -- the "causality" ordering
+                    // is preserved.
+                    var documentWriter = new BatchingLsifJsonWriter(_lsifJsonWriter);
+                    var documentId = await GenerateForDocumentAsync(document, compilation, project.LanguageServices, topLevelSymbolsResultSetTracker, documentWriter, _idFactory);
+                    topLevelSymbolsWriter.FlushToUnderlyingAndEmpty();
+                    documentWriter.FlushToUnderlyingAndEmpty();
+
+                    documentIds.Add(documentId);
+                }, CancellationToken.None));
+            }
+
+            Task.WaitAll(tasks.ToArrayAndFree());
```

Comment:
```suggestion
            Task.WaitAll(tasks);
```

Reply:
also, @mavasani @jasonmalinowski should we just make this an async prgram, so we can use async/await all the way up?

Reply:
@CyrusNajmabadi @mavasani I intentionally _didn't_ make it async, because if you want to use this in an analyzer it wasn't clear how async was going to work. Once you're at the compiler-only layer, there is no async to take advantage of, and any attempt might just block up more threads.